config: Make  gRPC xDS retry backoff configurable

jaysonjdmello opened the pull request:
Signed-off-by: Jayson Dmello <jdmello@confluent.io>

Commit Message:
The gRPC xDS implementation uses hardcoded retry durations for re-establishing the gRPC channel. This change makes the backoff parameters configurable.

Additional Description:
Risk Level: Low
Testing: unit test
Docs Changes:
Release Notes: Added
Platform Specific Features:
[Optional Runtime guard:]
 Fixes #24236 
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

KBaichoo commented:
/assign @yanjunxiang-google 

Can you take a look?

yanjunxiang-google commented on source/common/config/utility.h:
```diff
@@ -41,6 +41,9 @@ namespace Config {
 
 constexpr absl::string_view Wildcard = "*";
 
+static constexpr uint64_t RetryBaseIntervalMs = 500;
```
Are these global constants GRPC retry mechanism specific?  Need some comments here. Also is here the best place to define them?

jaysonjdmello replied:
These are not intended to be specific to GRPC retry timers but rather the the default time for @prepareBackoffStrategy method in this file.

yanjunxiang-google commented on test/common/config/utility_test.cc:
```diff
@@ -293,6 +293,155 @@ TEST(UtilityTest, PrepareDnsRefreshStrategy) {
+                          ->isOverTimeLimit(test_base_interval_ms * 10 + 1));
+    }
+
+    // set max_interval > base_interval
```
Do you mean max_interval < base_interval here?

jaysonjdmello replied:
will fix ,thanks a lot!

yanjunxiang-google commented:
LGTM in general. Just a few nit comments.

lizan commented on api/envoy/config/core/v3/grpc_service.proto:
```diff
@@ -288,10 +288,15 @@ message GrpcService {
   // documentation on :ref:`custom request headers
   // <config_http_conn_man_headers_custom_request_headers>`.
-  repeated HeaderValue initial_metadata = 5;
+  repeated HeaderValue initial_metadata = 6;
```
Don't renumber proto field, this breaks API backward compatibility.

jaysonjdmello replied:
corrected, thanks!

KBaichoo commented:
/assign @adisuissa 

For xDS code review maintainer pass

lizan commented:
/lgtm api

adisuissa reviewed (commented):
Thanks for your contribution.
Before going into the details, can you explain the benefit of using this over the [gRPC retries](https://github.com/grpc/proposal/blob/master/A6-client-retries.md) mechanism?

/wait-any

jaysonjdmello commented:
The gRPC xDS implementation uses hardcoded retry durations for re-establishing the gRPC channel. This change aims to make the retry durations configurable via envoy config. The purpose is to still use the gRPC retries mechanism as-is

adisuissa commented on api/envoy/config/core/v3/grpc_service.proto:
```diff
@@ -294,4 +294,9 @@ message GrpcService {
+  // Indicates the retry policy for re-establishing the gRPC channel
+  // This field is optional. If not provided, base interval is set to 500 ms, max interval set to 30 secs.
+  // If max interval is not provided, it will be set to ten times the base interval.
+  RetryPolicy retry_policy = 6;
```
The issue with adding this field here is that there are other places that use `GrpcService` (i.e., non-xDS) that at the moment do not parse this field.

I wonder if this should be a property of the underlying client, and then one just configures the retry policy for EnvoyGrpc, and uses the [gRPC retries](https://github.com/grpc/proposal/blob/master/A6-client-retries.md) mechanism for GoogleGrpc. @htuch any opinions on this?

jaysonjdmello replied:
Sorry, I am new to the project. 
Are you suggesting this is moved inside message EnvoyGrpc (https://github.com/envoyproxy/envoy/blob/bb7d941e7e245ebd89ca706d428d92f050d5e7b3/api/envoy/config/core/v3/grpc_service.proto#L32)?

adisuissa replied:
I think that EnvoyGrpc and GoogleGrpc will have different mechanism to handle backoff.
Specifically for this PR, I think you need to have the retry config in `EnvoyGrpc`, and the actual code that handles retries in [AsyncClientImpl](https://github.com/envoyproxy/envoy/blob/c394382139010828c125a23eccb47a5172e2eab8/source/common/grpc/async_client_impl.h#L23). An added value of this is that other uses of `AsyncClientImpl` (say filters that use gRPC), will also have that mechanism implemented.

The GoogleGrpc will have a different retry mechanism that is supported by the library.

jaysonjdmello replied:
Thanks a lot for the feedback.

 It looks like adding the backoff mechanism to `AsyncClientImpl` would require extensive design changes. The goal of https://github.com/envoyproxy/envoy/issues/24236 and this PR was to resolve the TODO of making the  backoff parameters configurable. 

I added the `BackoffStrategy` to `ApiConfigSource` for this iteration,  I can raise another issue for refactoring `EnvoyGrpc` retry mechanism if that works. 

Thanks again.

adisuissa replied:
Thanks! Will it be possible to add the config into `EnvoyGrpc`. The motivation is to avoid the need to deprecate the change to the `ApiConfigSource` in the future?

/wait-any

jaysonjdmello replied:
Sounds good, Moved the config into `EnvoyGrpc`.  Thanks

RyanTheOptimist commented:
/wait

jaysonjdmello commented:
/retest

jaysonjdmello commented:
/retest

jaysonjdmello commented:
/retest

adisuissa commented on source/common/config/utility.h:
```diff
@@ -531,6 +537,45 @@ class Utility {
+   * @return BackOffStrategyPtr
+   */
+  static BackOffStrategyPtr
+  prepareBackoffStrategy(const envoy::config::core::v3::ApiConfigSource& api_config_source,
```
The input is very specific to config-source, and it should probably be more generalized. Seems similar to the code [here](https://github.com/envoyproxy/envoy/blob/5991a2053e981038c105a421886af073be9c5941/source/common/config/datasource.cc#L70-L80).

jaysonjdmello replied:
done, refactored to a generalized implementation and reused in `source/common/config/datasource.cc`

adisuissa commented on envoy/common/backoff_strategy.h:
```diff
@@ -28,6 +28,14 @@ class BackOffStrategy {
+   * @param interval time interval to be checked.
+   * @return returns true if interval is greater than the maximum time interval
+   */
+  virtual bool isOverTimeLimit(uint64_t interval_ms) PURE;
```
The maximum time interval doesn't seem to be a shared property among all backoff strategies, but very specific to `JitteredExponentialBackOffStrategy`. I think this PR can be redesigned without introducing this to the BackoffStrategy interface.

jaysonjdmello replied:
done, redesigned the PR to use the derived class directly

adisuissa reviewed (changes requested):
Thanks for your contribution.
A high-level comment: there seems to be a mismatch between the expected use of a specific backoff strategy (`JitteredExponentialBackOffStrategy`), and using the generic type in GrpcStream. IIUC this comes from the need of GrpcStream to query the max-value of that derived class. I suggest using the derived class directly, instead of more generic backoff-strategy, as there is no way to change the type of strategy from the config.

/wait

jaysonjdmello commented on source/common/config/utility.h:
```diff
@@ -531,6 +537,45 @@ class Utility {
+  static BackOffStrategyPtr
+  prepareBackoffStrategy(const envoy::config::core::v3::ApiConfigSource& api_config_source,
+                         Random::RandomGenerator& random) {
+
```
moved implementation to `source/common/config/utility.cc`.

jaysonjdmello commented on api/envoy/config/core/v3/grpc_service.proto:
```diff
@@ -43,6 +44,11 @@ message GrpcService {
+
+    // Specifies parameters that control :ref:`retry backoff strategy <envoy_v3_api_msg_config.core.v3.BackoffStrategy>`.
+    // This parameter is optional, in which case the default base interval for retry is 500 milliseconds.
+    // The Default maximum interval is 30 seconds.
```
done

jaysonjdmello commented:
> 

Thanks for your feedback!
I have redesigned the PR to use the derived class directly. Also made other changes as suggested/

jaysonjdmello commented:
/retest

adisuissa commented on source/common/config/utility.h:
```diff
@@ -531,7 +531,23 @@ class Utility {
+   * @param random random generator
+   * @return JitteredExponentialBackOffStrategyPtr
+   */
+  static JitteredExponentialBackOffStrategyPtr prepareJitteredExponentialBackOffStrategy(
```
I think it would be better to have a 2 functions: one that receives the random-generator and an optional RetryPolicy (OptRef<const>), and another that is just like the first but with additional default values (the first function calling the second).
This way the function could check whether there is a retry policy, and if not, use the given default values.

jaysonjdmello replied:
Sorry, I am trying to confirm my understanding here. Are you referring to something like this?

 ```
static JitteredExponentialBackOffStrategyPtr prepareJitteredExponentialBackOffStrategy(
      OptRef<const envoy::config::core::v3::RetryPolicy> retry_policy,
      Random::RandomGenerator& random);
```
&

```
static JitteredExponentialBackOffStrategyPtr prepareJitteredExponentialBackOffStrategy(
      OptRef<const envoy::config::core::v3::RetryPolicy> retry_policy,
      Random::RandomGenerator& random, uint32_t retry_initial_delay_ms = 1000,
      uint32_t retry_max_delay_ms = 10000);
```

adisuissa reviewed (changes requested):
Thanks!
Added a few more comments that I hope will improve things.
/wait

jaysonjdmello commented on source/common/config/subscription_factory_impl.cc:
```diff
@@ -110,23 +125,35 @@ SubscriptionPtr SubscriptionFactoryImpl::subscriptionFromConfigSource(
+            grpc_services[0].envoy_grpc().retry_policy(), api_.randomGenerator());
+      } else {
+        backoff_strategy =
+            Utility::prepareDefaultJitteredExponentialBackOffStrategy(api_.randomGenerator());
```
agree. I think to reduce the impact of the change, I am proposing to not make changes to `datasource.cc` in this PR and to set the default values as is to :
```
constexpr uint32_t RetryInitialDelayMilliseconds = 500;
 constexpr uint32_t RetryMaxDelayMilliseconds = 30000;
```

jaysonjdmello replied:
> Thanks! Added a few more comments that I hope will improve things. /wait

Thank you! Updated with suggested comments and refactored the `prepareJitteredExponentialBackOffStrategy` to return `nullptr` if no back-off values are provided through the config.

jaysonjdmello commented on source/common/config/subscription_factory_impl.cc:
```diff
@@ -110,23 +125,35 @@ SubscriptionPtr SubscriptionFactoryImpl::subscriptionFromConfigSource(
+      } else {
+        backoff_strategy =
+            Utility::prepareDefaultJitteredExponentialBackOffStrategy(api_.randomGenerator());
+      }
```
Changed the function implementation and declaration

jaysonjdmello commented:
> Thanks! Added a few more comments that I hope will improve things. /wait

Thank a lot for your review! I  updated with suggested comments and refactored the `prepareJitteredExponentialBackOffStrategy` method

adisuissa commented on api/envoy/config/core/v3/grpc_service.proto:
```diff
@@ -43,6 +43,11 @@ message GrpcService {
+
+    // Indicates the retry policy for re-establishing the gRPC channel
+    // This field is optional. If max interval is not provided, it will be set to ten times the provided base interval.
+    // Currently only supported for xDS gRPC streams.
```
Please also state the defaults of the xDS gRPC streams.

jaysonjdmello replied:
done

adisuissa commented on source/common/config/datasource.cc:
```diff
@@ -64,24 +66,12 @@ RemoteAsyncDataProvider::RemoteAsyncDataProvider(
-        throw EnvoyException("max_interval must be greater than or equal to the base_interval");
-      }
-    }
+  backoff_strategy_ = Utility::prepareJitteredExponentialBackOffStrategy(source, random);
```
You can pass `RetryInitialDelayMilliseconds` and `RetryMaxDelayMilliseconds` as optionals to the `prepareJitteredExponentialBackOffStrategy` function, so it will always return a valid strategy, and will allow one to override the defaults.

jaysonjdmello replied:
done, thanks!

adisuissa reviewed (commented):
Thanks, and apologies for the late reply.
The review was pending for a while, and I didn't press the submit button.

jaysonjdmello commented on source/common/config/grpc_stream.h:
```diff
@@ -53,8 +54,10 @@ class GrpcStream : public Grpc::AsyncStreamCallbacks<ResponseProto>,
+    if (backoff_strategy_ == nullptr) {
+      backoff_strategy_ = std::make_unique<JitteredExponentialBackOffStrategy>(
+          RetryInitialDelayMs, RetryMaxDelayMs, random_);
+    }
```
done

jaysonjdmello commented on source/common/common/backoff_strategy.h:
```diff
@@ -34,6 +34,14 @@ class JitteredExponentialBackOffStrategy : public BackOffStrategy {
+   * @param interval time interval to be checked.
+   * @return returns true if interval is greater than the maximum time interval
+   */
+  bool isOverTimeLimit(uint64_t interval_ms) { return interval_ms > max_interval_; }
```
done

jaysonjdmello commented:
> Thanks, and apologies for the late reply. The review was pending for a while, and I didn't press the submit button.

Thank you for reviewing! I have addressed the review suggestions

adisuissa commented on source/common/config/subscription_factory_impl.cc:
```diff
@@ -77,6 +77,14 @@ SubscriptionPtr SubscriptionFactoryImpl::subscriptionFromConfigSource(
           std::make_unique<CustomConfigValidatorsImpl>(validation_visitor_, server_,
                                                        api_config_source.config_validators());
+
+      JitteredExponentialBackOffStrategyPtr backoff_strategy;
+      auto& grpc_services = api_config_source.grpc_services();
+      if (!grpc_services.empty() && grpc_services[0].has_envoy_grpc()) {
+        backoff_strategy = Utility::prepareJitteredExponentialBackOffStrategy(
+            grpc_services[0].envoy_grpc(), api_.randomGenerator(), absl::nullopt, absl::nullopt);
+      }
```
This code seems to be repeating itself a few times here.
I think it can be moved to an unnamed namespace function in this file, and also ensure that a backoff-strategy is always created (either with the given config, or default values). This should simplify the `GrpcStream` c'tor.

adisuissa replied:
EDIT: looks like this function repeats itself in another file (see below),so I think it should be static-public.

jaysonjdmello replied:
moved the repeated code to source/common/config/utility.h

adisuissa commented on source/common/config/utility.h:
```diff
@@ -531,7 +531,47 @@ class Utility {
+                                                       random);
+    }
+
+    // 2. Default base interval and Default maximum interval is specified
+    if (default_base_interval_ms != absl::nullopt && default_max_interval_ms != absl::nullopt) {
+      return std::make_unique<JitteredExponentialBackOffStrategy>(
+          default_base_interval_ms.value(), default_max_interval_ms.value(), random);
+    }
```
Looking at this again, I think this function can be further simplified. My goal is that this function will always return a valid strategy (non-nullptr), so its callers don't need to handle corner cases.
If the idea is that the caller will always pass a `default_base_interval_ms`, then the default max-interval will always be 10. WDYT?

jaysonjdmello replied:
the problem with always returning a valid strategy is that the default values in this case are in `grpc_stream.h`. but the config is accessible only in `subscription_factory_impl.cc`. 
We would have to move the defaults to `subscription_factory_impl.cc` but I am not sure if moving them from `grpc_stream.h`. is a good idea.

adisuissa replied:
> the problem with always returning a valid strategy is that the default values in this case are in `grpc_stream.h`. but the config is accessible only in `subscription_factory_impl.cc`. We would have to move the defaults to `subscription_factory_impl.cc` but I am not sure if moving them from `grpc_stream.h`. is a good idea.

Why should they stay in `grpc_stream.h`? I may be missing something, so feel free to explain a bit here.
FWIW, I think the [original PR](https://github.com/envoyproxy/envoy/pull/1124) had the default value in something that is more similar to the subscription factory that we have at the moment.

jaysonjdmello replied:
thanks for pointing out, I have moved the defaults to `envoy/config/subscription_factory.h`

adisuissa commented on source/common/upstream/cluster_manager_impl.cc:
```diff
@@ -384,6 +384,13 @@ ClusterManagerImpl::ClusterManagerImpl(
             validation_context.dynamicValidationVisitor(), server,
             dyn_resources.ads_config().config_validators());
 
+    JitteredExponentialBackOffStrategyPtr backoff_strategy;
+    auto& grpc_services = dyn_resources.ads_config().grpc_services();
+    if (!grpc_services.empty() && grpc_services[0].has_envoy_grpc()) {
+      backoff_strategy = Config::Utility::prepareJitteredExponentialBackOffStrategy(
+          grpc_services[0].envoy_grpc(), random_, absl::nullopt, absl::nullopt);
+    }
```
Looks like the function that I commented above.
In this case I think it should be a static function in the subscription-factory.

jaysonjdmello replied:
would it be better if this is moved to the `utility` class as a new method `prepareJitteredExponentialBackOffStrategy(const envoy::config::core::v3::ApiConfigSource& api_config_source, Random::RandomGenerator& random)` instead?

jaysonjdmello replied:
done, thanks!

adisuissa reviewed (commented):
Thanks, looks good.
Left a few minor comments.

markdroth commented on api/envoy/config/core/v3/grpc_service.proto:
```diff
@@ -43,6 +43,12 @@ message GrpcService {
         [(validate.rules).string =
              {min_len: 0 max_bytes: 16384 well_known_regex: HTTP_HEADER_VALUE strict: false}];
+
+    // Indicates the retry policy for re-establishing the gRPC channel
```
I think this comment is inaccurate.  If I'm reading the code correctly, this actually controls recreating the gRPC *stream*, not the gRPC channel.

jaysonjdmello replied:
agreed, will fix

markdroth commented on api/envoy/config/core/v3/grpc_service.proto:
```diff
@@ -43,6 +43,12 @@ message GrpcService {
+
+    // Indicates the retry policy for re-establishing the gRPC channel
+    // This field is optional. If max interval is not provided, it will be set to ten times the provided base interval.
+    // Currently only supported for xDS gRPC streams.
```
Why is this supported only for xDS streams?  Shouldn't this work for any stream?

If this is actually only for xDS streams, then it's not clear to me that it makes sense to put this in `GrpcService`, since that message is used for many things other than xDS streams.

jaysonjdmello replied:
The intent was to support this only for xDS streams in this iteration. We can add support for others in further iterations if that is okay.

jaysonjdmello commented:
> Thanks, looks good. Left a few minor comments.

Thanks for reviewing! refactored the changes

adisuissa commented on api/envoy/config/core/v3/grpc_service.proto:
```diff
@@ -43,6 +43,12 @@ message GrpcService {
+    // This field is optional. If max interval is not provided, it will be set to ten times the provided base interval.
+    // Currently only supported for xDS gRPC streams.
+    // If not set, xDS gRPC streams default base interval:500ms, maximum interval:30s will be applied.
+    RetryPolicy retry_policy = 3;
```
nit: just realized there are 2 RetryPolicy types, so let's make this explicit: `config.core.v3.RetryPolicy`.

jaysonjdmello replied:
done, thanks

jaysonjdmello replied:
reverted the change, it looks like the linter does not like it

```
INFO: Analyzed target @com_github_bufbuild_buf//:bin/buf (1 packages loaded, 1 target configured).
INFO: Found 1 target...
WARNING: /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/com_github_bufbuild_buf/BUILD.bazel:6:10: @com_github_bufbuild_buf//:bin/buf is a source file, nothing will be built for it. If you want to build a target that consumes this file, try --compile_one_dependency
INFO: Elapsed time: 0.938s, Critical Path: 0.04s
INFO: 1 process: 1 internal.
INFO: Running command line: /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/external/com_github_bufbuild_buf/bin/buf lint
INFO: Streaming build results to: https://source.cloud.google.com/results/invocations/6a1a3f8a-844c-42b2-a403-3c2bbbc1ca17
Format check failed, try apply following patch to fix:
diff --git a/api/envoy/config/core/v3/grpc_service.proto b/api/envoy/config/core/v3/grpc_service.proto
index 7cfe602..f266c7b 100644
--- a/api/envoy/config/core/v3/grpc_service.proto
+++ b/api/envoy/config/core/v3/grpc_service.proto
@@ -48,7 +48,7 @@ message GrpcService {
     // This field is optional. If max interval is not provided, it will be set to ten times the provided base interval.
     // Currently only supported for xDS gRPC streams.
     // If not set, xDS gRPC streams default base interval:500ms, maximum interval:30s will be applied.
-    config.core.v3.RetryPolicy retry_policy = 3;
+    RetryPolicy retry_policy = 3;
   }
 
   // [#next-free-field: 9]
```
I am guessing because `RetryPolicy` and `GrpcService` are part of the same package `envoy.config.core.v3`? Please correct me here, thanks

adisuissa replied:
Sorry about the back and forth.

jaysonjdmello replied:
No worries, thank you so much!

adisuissa commented on test/common/config/subscription_factory_impl_test.cc:
```diff
@@ -314,6 +314,34 @@ TEST_F(SubscriptionFactoryTest, HttpSubscriptionNoRefreshDelay) {
                             "refresh_delay is required for REST API configuration sources");
 }
 
+TEST_F(SubscriptionFactoryTest, GrpcSubscriptionInvalidRetryConfig) {
```
nit: please add a short comment before each test (here and in other tests) that explains what the test validates.

jaysonjdmello replied:
done

adisuissa commented on source/common/config/utility.cc:
```diff
@@ -308,5 +308,31 @@ void Utility::translateOpaqueConfig(const ProtobufWkt::Any& typed_config,
+    return std::make_unique<JitteredExponentialBackOffStrategy>(base_interval_ms, max_interval_ms,
+                                                                random);
+  }
+  // default maximum interval is specified
```
nit: Add an assertion constraint for the default_base_interval_ms (similar to the PGV constraint for `base_interval` in `BackoffStrategy`.

jaysonjdmello replied:
done, thanks

jaysonjdmello commented:
> Thanks! LGTM modulo small nits.

Thank you so much! refactored to address the review comments

jaysonjdmello commented:
/retest

jaysonjdmello commented:
/retest

adisuissa reviewed (approved):
Thanks, LGTM!

/assign-from @envoyproxy/senior-maintainers

htuch commented on envoy/config/subscription_factory.h:
```diff
@@ -13,6 +13,10 @@ class SubscriptionFactory {
 
+  // Default Parameters for the backoff strategy.
+  static constexpr uint32_t RetryInitialDelayMs = 500;
+  static constexpr uint32_t RetryMaxDelayMs = 30000; // Do not cross more than 30s
```
What is this comment intending to convey? If it's just "this is the max", then not needed (it's in the name). If there's some reason the max can't be more than 30s, it would be good to explain why.

jaysonjdmello replied:
The intent of the PR was to not change the legacy values and behavior from https://github.com/envoyproxy/envoy/blob/main/source/common/config/grpc_stream.h#L22

Unfortunately, I am not sure why 30s was chosen to be the maximum. I can remove the comment here if that is okay.

htuch replied:
I think just delete the comment, yes, seems a bit confusing.

htuch commented on test/common/config/utility_test.cc:
```diff
@@ -294,6 +294,235 @@ TEST(UtilityTest, PrepareDnsRefreshStrategy) {
+          test_max_interval_ms);
+
+      EXPECT_ANY_THROW(
+          Utility::prepareJitteredExponentialBackOffStrategy(config, random, 500, absl::nullopt));
```
Is there any test that actually test the retry behavior, rather than just that we build the correct configuration for the retry behavior?

htuch replied:
Well, the basic jitter backoff tests make sense as unit tests, but what I'm talking about here is some test that shows that we can both configure and observe this configuration take effect in a gRPC retry. Looking at this PR and `grpc_stream_test.cc`, I don't see any new tests or assertions covering this. FWIW you would need to use fake/mock time here to avoid flakes.

jaysonjdmello replied:
@htuch , added unit tests in `grpc_stream_test.cc` to test gRPC retry with custom retry values.

htuch commented:
/coverage

alyssawilk commented:
@htuch is this one waiting on anything?

htuch commented:
Yep, I think @jaysonjdmello was going to add some tests.
/wait

jaysonjdmello commented:
/retest

jaysonjdmello commented:
/retest

jaysonjdmello commented:
/retest

jaysonjdmello commented:
@htuch, would highly appreciate if you could provide feedback on the new added unit tests in `grpc_stream_test` , thanks!

htuch commented on source/common/config/utility.cc:
```diff
@@ -308,5 +308,41 @@ void Utility::translateOpaqueConfig(const ProtobufWkt::Any& typed_config,
+  }
+
+  // default_base_interval_ms must be greater than zero
+  if ((default_base_interval_ms > 0) == false) {
```
Nit: this is a weird way to write this expression, this should be `default_base_interval_ms == 0`.

jaysonjdmello replied:
done, thanks!

htuch commented on test/common/config/utility_test.cc:
```diff
@@ -294,6 +294,235 @@ TEST(UtilityTest, PrepareDnsRefreshStrategy) {
 }
 
+// test that default values are used correctly when no retry configuration is provided
+TEST(UtilityTest, prepareJitteredExponentialBackOffStrategyStrategyNoConfig) {
```
Nit: `PrepareJittered..`

jaysonjdmello replied:
done, thanks!

htuch commented on test/common/config/utility_test.cc:
```diff
@@ -294,6 +294,235 @@ TEST(UtilityTest, PrepareDnsRefreshStrategy) {
+
+// confirm that user provided values in the retry configuration are correctly used to prepare the
+// backoff strategy
+TEST(UtilityTest, prepareJitteredExponentialBackOffStrategyStrategyConfigFileValues) {
```
Nit: ditto

jaysonjdmello replied:
done, thanks!

htuch reviewed (commented):
LGTM modulo tiniest nits, thanks for the tests!

jaysonjdmello commented:
> LGTM modulo tiniest nits, thanks for the tests!

Thank you for reviewing! I have addressed the issues

jaysonjdmello commented:
/retest

htuch reviewed (approved):
LGTM, thanks!
